src/tpm2_wrap.c

```diff
@@ -3249,6 +3294,110 @@ int wolfTPM2_GetNvAttributesTemplate(TPM_HANDLE auth, word32* nvAttributes)
     return 0;
 }
 
+int wolfTPM2_CreateEK(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* ekKey, TPM_ALG_ID alg)
+{
+    int rc;
+    TPMT_PUBLIC publicTemplate;
+
+    if(alg == TPM_ALG_RSA) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_EK(&publicTemplate);
+    }
+    else if(alg == TPM_ALG_ECC) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_EK(&publicTemplate);
+    }
+    else {
+        /* Supported algorithms for EK are only 2048bit RSA & ECC */
+        return BAD_FUNC_ARG;
+    }
+    /* GetKeyTemplate check */
+    if (rc != 0) return rc;
+
+    rc = wolfTPM2_CreatePrimaryKey(dev, ekKey, TPM_RH_ENDORSEMENT,
+        &publicTemplate, NULL, 0);
+
+    return rc;
+}
+
+int wolfTPM2_CreateSRK(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* srkKey, TPM_ALG_ID alg,
+    const byte* auth, int authSz)
+{
+    int rc;
+    TPMT_PUBLIC publicTemplate;
+
+    /* Supported algorithms for SRK are only 2048bit RSA & ECC */
+    if(alg == TPM_ALG_RSA) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_SRK(&publicTemplate);
+    }
+    else if(alg == TPM_ALG_ECC) {
+        /* TODO: Create GetKeyTemplate_ECC_SRK helper */
+        return BAD_FUNC_ARG;
+    }
+    else {
+        /* Supported algorithms for EK are only 2048bit RSA & ECC */
+        return BAD_FUNC_ARG;
+    }
+    /* GetKeyTemplate check */
+    if (rc != 0) return rc;
+
+    rc = wolfTPM2_CreatePrimaryKey(dev, srkKey, TPM_RH_OWNER,
+        &publicTemplate, auth, authSz);
+
+    return rc;
+}
+
+int wolfTPM2_CreateAndLoadAIK(WOLFTPM2_DEV* dev, WOLFTPM2_KEY* aikKey,
+    TPM_ALG_ID alg, WOLFTPM2_KEY* srkKey, const byte* auth, int authSz)
+{
+    int rc;
+    TPMT_PUBLIC publicTemplate;
+
+    if(alg == TPM_ALG_RSA) {
+        rc = wolfTPM2_GetKeyTemplate_RSA_AIK(&publicTemplate);
+    }
+    else {
+        return BAD_FUNC_ARG;
+    }
+    /* GetKeyTemplate check */
+    if (rc != 0) return rc;
+
+    rc = wolfTPM2_CreateAndLoadKey(dev, aikKey, &srkKey->handle,
+        &publicTemplate, auth, authSz);
+
+    return rc;
+}
+
+int wolfTPM2_GetTime(WOLFTPM2_KEY* aikKey, GetTime_Out* out)
+{
+    int rc;
+    GetTime_In getTimeCmd;
+    GetTime_Out getTimeOut;
+
+    if(out == NULL) return BAD_FUNC_ARG;
+
+    /* GetTime */
+    XMEMSET(&getTimeCmd, 0, sizeof(getTimeCmd));
+    XMEMSET(&getTimeOut, 0, sizeof(getTimeOut));
+    getTimeCmd.privacyAdminHandle = TPM_RH_ENDORSEMENT;
+    /* TPM_RH_NULL is a valid handle for NULL signature */
+    getTimeCmd.signHandle = aikKey->handle.hndl;
+    /* TPM_ALG_NULL is a valid handle for  NULL signature */
+    getTimeCmd.inScheme.scheme = TPM_ALG_RSASSA;
+    getTimeCmd.inScheme.details.rsassa.hashAlg = TPM_ALG_SHA256;
+    getTimeCmd.qualifyingData.size = 0; /* optional */
+    rc = TPM2_GetTime(&getTimeCmd, &getTimeOut);
+    if (rc != TPM_RC_SUCCESS) {
+    #ifdef DEBUG_WOLFTPM
+        printf("TPM2_GetTime failed 0x%x: %s\n", rc, wolfTPM2_GetRCString(rc));
+    #endif
+    }
+
+    /* TODO: Decide what to return. For now, the output structure of GetTime */
+    out = &getTimeOut;
```

Comment:
Returning pointer to stack variable is invalid. You should use `out` directly in the `TPM2_GetTime` call and delete `getTimeOut`.

Reply:
Indeed. Thanks